File: Memento/Dart/EditorHistoryCaretaker.dart
import 'WorkspaceState.dart';

class EditorHistoryCaretaker {
  List<WorkspaceState> _states = [];

  void add(WorkspaceState? state) {
    if (state != null) {
      _states.add(state);
    }
  }

  WorkspaceState? get(int index) {
    if (index < 0 || _states.length < 1) {
      return null;
    }

    if (_states.length <= index) {
      return _states[_states.length - 1];
    } else {
      return _states[index];
    }
  }
}

File: Memento/Javascript/EditorHistoryCaretaker.js
import WorkspaceState from "./WorkspaceState.js";

class EditorHistoryCaretaker {
  constructor() {
    this.__states = [];
  }

  add(state) {
    this.__states.push(state);
  }

  get(index) {
    if (this.__states.length < 1) {
      return null;
    }

    if (index < this.__states.length) {
      return this.__states[index];
    } else {
      return this.__states[this.__states.length - 1];
    }
  }
}

export default EditorHistoryCaretaker;